### PR TITLE
Refine LoadBalancer service access control and consolidating security-rules

### LINKED ISSUE #4713: LoadBalancerSourceRanges: Multiple CIDRs for one NSG rule
### What would you like to be added:

Currently, if you set `spec.loadBalancerSourceRanges` to use multiple CIDRs, it will create 1 network security group (NSG) rule per CIDR. We need a mechanism in which we could specify `spec.loadBalancerSourceRanges` for 1 NSG rule with multiple CIDRs. 

### Why is this needed:

We need more than 1000 CIDRs. The [1000 NSG rule limit](https://github.com/MicrosoftDocs/azure-docs/blob/main/includes/azure-virtual-network-limits.md) also results in a limitation of 1000 CIDRs since one NSG rule is created for each CIDR. By associating the CIDRs with one NSG rule, the limit of CIDRS would effectively be increased to 4000.

### Design Ideas:

1. **Create rules with multiple CIDRs**

    This update involves transitioning from the creation of multiple NSG rules to the creation of a single rule with multiple CIDRs. Are there any drawbacks to combining rules, e.g. performance, limits, or management concerns? If there is some advantage to using separate rules, we could add an option for combining rules, such as a service annotation.

~~2. **Enable use of pre-created NSGs**~~

~~With this approach, the cluster admin would need to precreate the desired NSGs.  However, this approach would require another API field, such as a service annotation like AWS's [service.beta.kubernetes.io/aws-load-balancer-security-groups](https://aws.amazon.com/blogs/containers/network-load-balancers-now-support-security-groups/) service annotation.~~ **Edit**: I'm removing this idea as there was some confusion about NSGs. As far as I understand, NSGs are linked to subnets, and it seems illogical to me to provide your own NSG alongside a Service, as it would need to apply to the subnets, not the load balancer.

### LINKED ISSUE #4919: Shared NSG rule should be deleted if the rule is not referred.
#### What happened:
Shared NSG rule should be deleted if the rule is not referred. 
#### What you expected to happen:
Shared NSG rule were not deleted if the svc was deleted. 
#### Environment:

- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by lodrem**
#### What type of PR is this?

/cleanup

#### What this PR does / why we need it:

Purpose: create as few security rules as possible. (detail in #4713)
This PR has a ton of changes because we’ve completely changed the pattern for security rules.

Some screenshots:

##### Without allow list

``` yaml
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  type: LoadBalancer
  ipFamilyPolicy: PreferDualStack
  selector:
    app: foo
  ports:
    - name: port1
      protocol: TCP
      port: 30102
      targetPort: 30102
    - name: port2
      protocol: UDP
      port: 30102
      targetPort: 30102
    - name: port3
      protocol: TCP
      port: 30103
      targetPort: 30103
```
IPv4
image

IPv4 and IPv6
image

##### With additional public IPs
```yaml
apiVersion: v1
kind: Service
metadata:
  name: foo
  annotations:
    service.beta.kubernetes.io/azure-additional-public-ips: "10.0.0.1,10.0.0.2"
spec:
  type: LoadBalancer
  ipFamilyPolicy: PreferDualStack
  selector:
    app: foo
  ports:
    - name: port1
      protocol: TCP
      port: 30102
      targetPort: 30102
    - name: port2
      protocol: UDP
      port: 30102
      targetPort: 30102
    - name: port3
      protocol: TCP
      port: 30103
      targetPort: 30103
```
image

##### Disable floating IP
``` yaml
apiVersion: v1
kind: Service
metadata:
  name: foo
  annotations:
    service.beta.kubernetes.io/azure-disable-load-balancer-floating-ip: "true"
spec:
  type: LoadBalancer
  ipFamilyPolicy: PreferDualStack
  selector:
    app: foo
  ports:
    - name: port1
      protocol: TCP
      port: 30102
      targetPort: 30102
    - name: port2
      protocol: UDP
      port: 30102
      targetPort: 30102
    - name: port3
      protocol: TCP
      port: 30103
      targetPort: 30103
```
NOTE: use backend pool nodes IPs as the destination addresses
image

##### With allow list

``` yaml
apiVersion: v1
kind: Service
metadata:
  name: foo
  annotations:
    service.beta.kubernetes.io/azure-allowed-service-tags: "AzureCloud,AzureDatabricks"
    service.beta.kubernetes.io/azure-allowed-ip-ranges: "10.10.10.10/32,10.10.10.11/32,2001:1234::/64"
spec:
  type: LoadBalancer
  ipFamilyPolicy: PreferDualStack
  selector:
    app: foo
  ports:
    - name: port1
      protocol: TCP
      port: 30102
      targetPort: 30102
    - name: port2
      protocol: UDP
      port: 30102
      targetPort: 30102
    - name: port3
      protocol: TCP
      port: 30103
      targetPort: 30103
```
image

#### Which issue(s) this PR fixes:

Fixes #4713
Fixes #4919 

#### Special notes for your reviewer:

I’ve rewritten the logic that handles the configuration, but it should still behave the same way:
- `service.beta.kubernetes.io/azure-allowed-ip-ranges`
- `service.beta.kubernetes.io/azure-allowed-service-tags`
- `service.beta.kubernetes.io/azure-disable-load-balancer-floating-ip`
- `service.beta.kubernetes.io/azure-additional-public-ips`
- and spec `spec.LoadBalancerSourceRanges`

The annotations would be deprecated after this change:
- `service.beta.kubernetes.io/azure-shared-securityrule` now all the rules would be as shared.

**code review guidance**
This PR is pretty big, so I’ll break it down here to make it less of a headache. Sorry in advance!
- Essentially, the logic for access control and security groups is in [./pkg/provider/loadbalancer/..](https://github.com/lodrem/cloud-provider-azure/tree/refactor-lb/consolidate-security-rules/pkg/provider/loadbalancer). Being stateless, it doesn’t involve any outgoing HTTP or similar requests, nor does it use memory cache or global variables.
- The `SecurityGroup` put operation would still be in [./pkg/provider/azure_loadbalancer.go#reconcileSecurityGroup](https://github.com/lodrem/cloud-provider-azure/blob/dc95d8b4c96cdf1f3554d0b7d565f452567a1736/pkg/provider/azure_loadbalancer.go#L2930)
- The unit tests have been rewritten, you can find them in [./pkg/provider/azure_loadbalancer_accesscontrol_test.go](https://github.com/lodrem/cloud-provider-azure/blob/refactor-lb/consolidate-security-rules/pkg/provider/azure_loadbalancer_accesscontrol_test.go)

The E2E tests aren’t finished yet (they still need some refactoring), but you can already start reviewing.

#### Does this PR introduce a user-facing change?

```release-note
Refine consolidating security-rules for LoadBalancer service.
Deprecate service annotation `service.beta.kubernetes.io/azure-shared-securityrule`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by lodrem** [comment]
> The IPV6 address in the last example does not appear in the source column of the screenshot. Is it expected?

Yes because that cluster didn't enable IPv6.

If dual stack enabled, you would see something like:
allowlist-dualstack

File: pkg/provider/azure_loadbalancer.go [hunk 1]
```diff
@@ -42,12 +41,14 @@ import (
 	servicehelpers "k8s.io/cloud-provider/service/helpers"
 	"k8s.io/klog/v2"
 	"k8s.io/utils/pointer"
+	"k8s.io/utils/ptr"
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer.go:44]: Can we use k8s.io/utils/pointer?

Comment (lodrem) [reply pkg/provider/azure_loadbalancer.go:44]: It looks like the `k8s.io/utils/pointer` is just a wrapper for `k8s.io/utils/ptr`. Plus, the package had been deprecated and we should use ptr instead: https://github.com/kubernetes/utils/blob/master/pointer/pointer.go#L17-L19.

File: pkg/provider/loadbalancer/accesscontrol.go [hunk 2]
```diff
@@ -21,182 +21,302 @@ import (
...
 	}
 	// Internal LB with explicit allowedAll IP ranges is allowed to be accessed from internet.
-	return len(ac.allowedIPRanges) > 0 || len(ac.sourceRanges) > 0
+	return len(ac.AllowedIPRanges) > 0 || len(ac.SourceRanges) > 0
 }
 
-// IPV4Sources returns the allowed sources for IPv4.
-func (ac *AccessControl) IPV4Sources() []string {
-	var rv []string
+// DenyAllExceptSourceRanges returns ture if it needs to block any VNet traffic not on the allow list.
```
Comment (feiskyer) [inline pkg/provider/loadbalancer/accesscontrol.go:142]: s/ture/true

File: pkg/provider/loadbalancer/securitygroup/securitygroup.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+	helper.rules[name] = rule
+	helper.priorities[priority] = name
+
+	logger.V(4).Info("Adding a new rule", "rule-name", name, "priority", priority)
+
+	return rule, nil
+}
+
+func (helper *RuleHelper) addAllowRule(
```
Comment (feiskyer) [inline pkg/provider/loadbalancer/securitygroup/securitygroup.go:163]: Could you add comments for newly introduced functions and packages?

File: pkg/provider/azure_loadbalancer_accesscontrol_test.go [hunk 4]
```diff
@@ -0,0 +1,2357 @@
...
+	)
+
+	var (
+		fx      = fixture.NewFixture()
+		k8sFx   = fx.Kubernetes()
+		azureFx = fx.Azure()
+	)
+
+	t.Run("internal Load Balancer", func(t *testing.T) {
+		t.Run("noop when no allow list specified", func(t *testing.T) {
```
Comment (feiskyer) [inline pkg/provider/azure_loadbalancer_accesscontrol_test.go:55]: any tests when ILB services are annotated with allowded CIRDs and service tags?

**Review by feiskyer — COMMENTED** [review]
Thank you for the excellent refactoring. The code is much clearer now. The changes LGTM in general, just added a few comments.

File: pkg/provider/azure_loadbalancer_accesscontrol_test.go (hunk 4 above)
Comment (lodrem) [reply pkg/provider/azure_loadbalancer_accesscontrol_test.go:55]: Good catch. Added a few more tests for ILB. They should cover its code path.

File: pkg/provider/loadbalancer/securitygroup/securitygroup.go (hunk 3 above)
Comment (lodrem) [reply pkg/provider/loadbalancer/securitygroup/securitygroup.go:168]: Added.

File: pkg/provider/loadbalancer/accesscontrol.go (hunk 2 above)
Comment (lodrem) [reply pkg/provider/loadbalancer/accesscontrol.go]: Fixed!

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Comment by phealy** [comment]
Looks amazing - thank you for all the hard work and reviews, @lodrem , @nilo19 , @feiskyer, @lzhecheng, and anyone else I missed!

**Pull request merged**